plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java

```diff
@@ -149,34 +156,32 @@ private void checkInputReady()
 
     private synchronized List<ExchangeSourceHandle> createExchangeSourceHandles()
     {
-        Multimap<Integer, Path> partitionFilesMap = ArrayListMultimap.create();
-        for (LocalFileSystemExchangeSinkHandle sinkHandle : finishedSinks) {
-            Path committedAttemptPath = getCommittedAttemptPath(sinkHandle);
-            Map<Integer, Path> partitions = getCommittedPartitions(committedAttemptPath);
+        Multimap<Integer, URI> partitionFilesMap = ArrayListMultimap.create();
+        for (Integer partitionId : finishedPartitions) {
+            URI committedAttemptPath = getCommittedAttemptPath(partitionId);
+            Map<Integer, URI> partitions = getCommittedPartitions(committedAttemptPath);
             partitions.forEach(partitionFilesMap::put);
         }
 
         ImmutableList.Builder<ExchangeSourceHandle> result = ImmutableList.builder();
         for (Integer partitionId : partitionFilesMap.keySet()) {
-            result.add(new LocalFileSystemExchangeSourceHandle(partitionId, ImmutableList.copyOf(partitionFilesMap.get(partitionId))));
+            result.add(new FileSystemExchangeSourceHandle(partitionId, ImmutableList.copyOf(partitionFilesMap.get(partitionId)), secretKey));
         }
         return result.build();
     }
 
-    private Path getCommittedAttemptPath(LocalFileSystemExchangeSinkHandle sinkHandle)
+    private URI getCommittedAttemptPath(int partitionId)
     {
-        Path sinkOutputBasePath = getExchangeDirectory()
-                .resolve(Integer.toString(sinkHandle.getTaskPartitionId()));
+        URI sinkOutputBasePath = getExchangeDirectory()
+                .resolve(partitionId + PATH_SEPARATOR);
         try {
-            List<Path> attemptPaths = Files.list(sinkOutputBasePath)
-                    .filter(Files::isDirectory)
-                    .collect(toImmutableList());
-            checkState(!attemptPaths.isEmpty(), "no attempts found for sink %s", sinkHandle);
+            List<URI> attemptPaths = exchangeStorage.listDirectories(sinkOutputBasePath).collect(toImmutableList());
+            checkState(!attemptPaths.isEmpty(), "no attempts found under sink output path %s", sinkOutputBasePath);
 
-            List<Path> committedAttemptPaths = attemptPaths.stream()
-                    .filter(LocalFileSystemExchange::isCommitted)
+            List<URI> committedAttemptPaths = attemptPaths.stream()
+                    .filter(this::isCommitted)
                     .collect(toImmutableList());
-            checkState(!committedAttemptPaths.isEmpty(), "no committed attempts found for %s", sinkHandle);
+            checkState(!committedAttemptPaths.isEmpty(), "no committed attempts found under sink output path %s", sinkOutputBasePath);
```

Comment:
```java
            return attemptPaths.stream()
                    .filter(this::isCommitted)
                    .findFirst()
                    .orElseThrow(() -> new IllegalStateException(format("no committed attempts found under sink output path %s", sinkOutputBasePath)));

```

Reply:
Actually I removed this check. I realized that this may not hold in case of a union, so to construct `ExchangeSource`, each file has a corresponding secret key.